ARROW-4503: [C#] Eliminate allocations in ArrowStreamReader when reading from a Stream

eerhardt opened the pull request:
When reading `RecordBatch` instances from a .NET `Stream` using the `ArrowStreamReader` class, it is currently allocating and copying memory 3 times for the data.

I've changed it to take a MemoryPool in the constructor, `ArrowStreamReader` will allocate from the pool a single time for each `RecordBatch` it reads, copy from the `Stream` into that memory once, and then use it in the returned `RecordBatch`.

I've also fixed the comments from #3736 that were deemed to be better fixed with this change.

eerhardt commented:
Here are some preliminary perf results of this change:

|                      Method |      Mean |     Error |    Gen 0/1k Op | Allocated Memory/Op |
|---------------------------- |----------:|----------:|----------:|--------------------:|
| **BEFORE** ArrowReaderWithMemoryStream | 81.417 ms | 1.6083 ms | 3000.0000 |         84000.73 KB |
| **AFTER** ArrowReaderWithMemoryStream | 29.431 ms | 0.5826 ms |             - |             6.28 KB |

The reason `Allocated Memory/Op` drops to almost 0 is because we are allocating from the default `MemoryPool`, which pulls from native memory, which isn't tracked by BenchmarkDotNet.

chutchinson commented:
@eerhardt Is there a way to report the native memory to the benchmark? Seems like it would be valuable for benchmarks, especially considering the current allocator is naive and probably even broken and/or not thread-safe. I expect it will be replaced or supplemented with Jemalloc.NET or some other more reliable allocation mechanism.

eerhardt commented:
> Is there a way to report the native memory to the benchmark?

I'm not sure, but maybe @adamsitnik might be able to answer.

On a related topic, @chutchinson, is there a specific reason why the default MemoryPool uses native memory? I guess my assumption was that the default should just be regular `byte[]` instances, which could be garbage collected. I also noticed that there doesn't seem to be a way to return memory back to the pool to be reused. Is that intentional?

eerhardt commented:
@stephentoub @chutchinson @pgovind - this is now ready for review.

stephentoub commented on csharp/src/Apache.Arrow/Extensions/StreamExtensions.netcoreapp2.1.cs:
```diff
@@ -0,0 +1,29 @@
+    // Helpers to read from Stream to Memory<byte> on netcoreapp
+    internal static class StreamExtensions
+    {
+        public static int Read(this Stream stream, Memory<byte> buffer)
```
Why is this necessary?  Why not just call it with buffer.Span at the call site?

stephentoub replied:
I see. You've added this as an abstraction to enable fishing out the array when the span-based overload doesn't exist on netstandard. Ok.

eerhardt replied:
Is there are better way to do this? I didn't find one.

stephentoub replied:
Given the desire to also compile for netstandard, this seems reasonable.

stephentoub commented on csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs:
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
 
-                bytesRead = await BaseStream.ReadAsync(buffer, 0, 4).ConfigureAwait(false);
-                footerLength = BinaryPrimitives.ReadInt32LittleEndian(buffer);
+                bytesRead = await BaseStream.ReadAsync(buffer.AsMemory(0, 4)).ConfigureAwait(false);
```
You'd be better off moving the declaration of bytesRead into each of these delegates so they don't have to close over the shared field and instead it can just be a local (that doesn't need to be lifted to any state machine).

eerhardt replied:
Good call. Fixed.

stephentoub commented on csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs:
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
-                    $"Failed to read footer length. Read <{bytesRead}>, expected 4.");
+            await Buffers.RentReturnAsync(footerLength, async (buffer) =>
+            {
+                _footerStartPostion = (int)BaseStream.Length - footerLength - ArrowFileConstants.Magic.Length - 4;
```
Is buffer.Length == footerLength?  If so, you could use that instead of footerLength to avoid closing over footerLength.  If instead buffer is a byte[] rented from the pool and thus could have a length > footerLength, you might change the signature to pass in a `Memory<byte>` rather than a `byte[]`, with it sliced to the right length, and then you could use `buffer.Length` for that purpose.

stephentoub commented on csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs:
```diff
@@ -26,12 +27,14 @@ internal class ArrowStreamReaderImplementation : ArrowReaderImplementation
+            _memoryPool = memoryPool ?? MemoryPool.Default.Value;
             _leaveOpen = leaveOpen;
-            Buffers = ArrayPool<byte>.Create();
+            Buffers = ArrayPool<byte>.Shared;
```
Does it need to be a property?  It'd be more efficient to just have all callers use `ArrayPool<byte>.Shared.Rent/Return`, both because it would remove the extra field from this instance and because the .NET Core JIT will likely be able to devirtualize those calls.

eerhardt replied:
Fixed.

stephentoub commented on csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs:
```diff
@@ -51,69 +54,101 @@ public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationTok
 
-                lengthBuffer = Buffers.Rent(4);
-                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken)
+                Memory<byte> bodyBuff = _memoryPool.Allocate((int)message.BodyLength);
```
Where and when do these buffers get returned to the pool?

eerhardt replied:
I have changed this to return an `IMemoryOwner<byte>` which is then wrapped by `IDisposable` objects. When those objects (`RecordBatch`, `ArrowBuffer`) get disposed, they dispose the `IMemoryOwner`.

stephentoub commented:
> ArrowReaderWithMemoryStream

Do we expect this to be a common case? If so, have you considered extracting the byte[] underlying the MemoryStream and using the memory-based reader instead of stream-based? Your need to read the header info at least presumably to know how to slice it, but that should then should then allow that case to be zero-copy, too.

chutchinson commented:
> On a related topic, @chutchinson, is there a specific reason why the default MemoryPool uses native memory? I guess my assumption was that the default should just be regular `byte[]` instances, which could be garbage collected. I also noticed that there doesn't seem to be a way to return memory back to the pool to be reused. Is that intentional?

The default memory pool allocates native memory because the allocated pointer can be adjusted to satisfy the Arrow memory alignment rule in a cross-platform way without GC pinning and the associated complexity. Admittedly I don't have a sufficiently deep understanding of the CLR (especially CoreCLR) to imagine a better way. 

There needs to be more thought and care put into the design of the memory allocation API and implementation. There is no mechanism for returning allocated memory; at the moment it is garbage collected as appropriate via the MemoryManager<T> API.

stephentoub commented:
> There is no mechanism for returning allocated memory; at the moment it is garbage collected as appropriate via the MemoryManager API.

Given this, I'd recommend not using this pool in this PR.  It's going to make things significantly less efficient.

bkietz commented on csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs:
```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
+
+            Buffers.RentReturn(4, (buffer) =>
+            {
+                BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
```
`BaseStream.Length - ArrowFileConstants.Magic.Length - 4` appears several times, could it be extracted into a well named method?

eerhardt replied:
This was fixed in #4017.

adamsitnik commented:
> Is there a way to report the native memory to the benchmark?

eerhardt commented:
OK - I finally got back to this PR and it is ready for re-review. I believe I have addressed all feedback and concerns. Please let me know.

Here are some updated numbers with the new benchmarks added:

## Before
|                      Method |   Count |         Mean |        Error |      StdDev | Gen 0/1k Op | Allocated Memory/Op |
|---------------------------- |-------- |-------------:|-------------:|------------:|------------:|--------------------:|
| ArrowReaderWithMemoryStream |   10000 |    639.69 us |    25.797 us |    71.91 us |   3000.0000 |           947.22 KB |
|       ArrowReaderWithMemory |   10000 |     59.98 us |     6.349 us |    17.70 us |           - |             4.48 KB |
| ArrowReaderWithMemoryStream | 1000000 | 78,008.70 us | 1,536.342 us | 1,437.10 us |   3000.0000 |         83999.47 KB |
|       ArrowReaderWithMemory | 1000000 |  2,675.14 us |   119.735 us |   333.77 us |           - |             4.48 KB |

## After
|                                    Method |   Count |         Mean |      Error |    StdDev | Gen 0/1k Op |  Allocated Memory/Op |
|------------------------------------------ |-------- |-------------:|-----------:|----------:|------------:|--------------------:|
|               ArrowReaderWithMemoryStream |   10000 |    185.59 us |  22.262 us |  65.29 us |   1000.0000 |              5.2 KB |
| ArrowReaderWithMemoryStream_ManagedMemory |   10000 |    110.95 us |  11.195 us |  32.12 us |           - |             5.19 KB |
|                     ArrowReaderWithMemory |   10000 |     55.50 us |   5.738 us |  16.37 us |           - |              4.7 KB |
|               ArrowReaderWithMemoryStream | 1000000 | 34,292.18 us | 582.247 us | 544.63 us |   1000.0000 |              5.2 KB |
| ArrowReaderWithMemoryStream_ManagedMemory | 1000000 | 27,583.96 us | 600.099 us | 616.26 us |           - |          41997.4 KB |
|                     ArrowReaderWithMemory | 1000000 |  2,897.50 us | 157.003 us | 462.93 us |           - |              4.7 KB |

Notes on the benchmarks:
1. I couldn't do the `_ManagedMemory` tests before this change, since that API didn't exist.
2. Notice the gains on `ArrowReaderWithMemoryStream_ManagedMemory` with `Count=10,000`. It is a 40% gain vs. the default case. The reasoning is because the `_ManagedMemory` uses `MemoryPool<byte>` to allocate managed memory. In the case of `10,000` records (the default in Spark), we are able to pool the memory buffers and don't need to allocate a new buffer each iteration.

adamsitnik commented on csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs:
```diff
@@ -78,23 +79,24 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
-                    EnsureFullRead(bodyBuff, bodyBytesRead);
+                int bodyLength = checked((int)message.BodyLength);
+
+                IMemoryOwner<byte> bodyBuffOwner = _allocator.Allocate(bodyLength);
```
again: is it acceptable to have `bodyBuffOwner` be a null?

what is the outcome of reading into an empty memory?

eerhardt replied:
What else would you expect to be returned when `0` gets passed into `Allocate`? Is allocating an object with an empty Memory worth it?

adamsitnik replied:
I asked this question because I did not understand why it's returning a null, not because I thought it's wrong to do so.

adamsitnik commented on csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs:
```diff
@@ -78,23 +79,24 @@ protected async ValueTask<RecordBatch> ReadRecordBatchAsync(CancellationToken ca
-                    result = CreateArrowObjectFromMessage(message, bodybb);
-                }).ConfigureAwait(false);
+                FlatBuffers.ByteBuffer bodybb = CreateByteBuffer(bodyBuff);
+                result = CreateArrowObjectFromMessage(message, bodybb, bodyBuffOwner);
```
should the owner be disposed here?

edit: I can see that `RecordBatch` is disposable now. Is the Arrow deserialization lazy and we need to keep the memory alive until somebody reads from it?

eerhardt replied:
The deserialization happens during `CreateArrowObjectFromMessage`. That method creates a `RecordBatch` and associated `IArrowArray` objects pointing into the memory here. The memory needs to be kept alive because how else would we remember all the values that we just read from the Stream?

adamsitnik replied:
It makes sense to me now #solved

adamsitnik commented on csharp/src/Apache.Arrow/RecordBatch.cs:
```diff
@@ -40,6 +42,24 @@ public IArrowArray Column(string columnName)
             return _arrays[fieldIndex];
         }
 
+        public void Dispose()
```
are there any types that derive from `RecordBatch`? Do we need full dispose pattern implementation here?

eerhardt replied:
It isn't sealed, and it has a `public` constructor, so I felt it was best. If we seal the class, we can remove the full dispose pattern.

adamsitnik commented on csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs:
```diff
@@ -47,6 +48,55 @@ public void Ctor_LeaveOpenTrue_StreamValidOnDispose()
+                (stream, memoryPool) => new ArrowFileReader(stream, memoryPool));
+        }
+
+        [Fact]
```
nit: you could use theory here and add two test cases: true & false

eerhardt replied:
Good call. Fixed.

adamsitnik commented on csharp/test/Apache.Arrow.Tests/ArrowFileReaderTests.cs:
```diff
@@ -47,6 +48,55 @@ public void Ctor_LeaveOpenTrue_StreamValidOnDispose()
+        [Fact]
+        public async Task Ctor_MemoryPool_AllocatesFromPool()
+        {
+            await Ctor_MemoryPool_AllocatesFromPoolHelper(
```
nit: you can just return the task without awaiting it (the unit test runner should do it)

eerhardt replied:
With the above theory, I eliminated this helper method all together.

pgovind commented on csharp/src/Apache.Arrow/ArrowBuffer.cs:
```diff
@@ -14,21 +14,37 @@
+
         public static ArrowBuffer Empty => new ArrowBuffer(Memory<byte>.Empty);
 
         internal ArrowBuffer(ReadOnlyMemory<byte> data)
```
Perhaps make this constructor public?

eerhardt replied:
I thought you were doing that in https://github.com/apache/arrow/pull/3971?  (BTW - I missed that you updated that PR, and approved it.)

chutchinson commented on csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs:
```diff
@@ -41,14 +42,10 @@ protected override Memory<byte> AllocateInternal(int length, out int bytesAlloca
 
-            return manager.Memory;
-        }
+            // Ensure all allocated memory is zeroed.
```
> why do you zero the memory? I would expect that it might have some random values, but they are getting overwritten anyway by the reader later.

I was responsible for this. It was useful during development to identify padding when inspecting buffers (since all padding is zeroed), and at the time we desired the same memory allocation semantics as general .NET primitives. I still see the value in zeroing the memory, but perhaps it can be placed behind a flag and defaulted to false for a performance advantage (if any)?

eerhardt replied:
I was just moving this zeroing from the base class (which would happen for ANY allocator) to just this native memory case. This way a managed memory implementation doesn't need to re-zero the memory, if it doesn't want to. Can you open a JIRA for this?

adamsitnik replied:
Thanks for the explanation. If it shows up in the profiles in the next performance investigation we can consider removing it then. #resolved

eerhardt replied:
I've opened https://issues.apache.org/jira/browse/ARROW-5276 to add this option.

eerhardt commented:
> I also wonder if we could somehow force the users to Dispose the Batch properly?

It is `IDisposable`, I would assume any analyzer that enforced `IDisposable` should work. e.g. https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2213-disposable-fields-should-be-disposed

eerhardt commented on csharp/src/Apache.Arrow/Arrays/ArrayData.cs:
```diff
@@ -53,5 +54,24 @@ public sealed class ArrayData
+            {
+                foreach (ArrayData child in Children)
+                {
+                    child?.Dispose();
```
I don't see any code that guards against having a `null` in the `Children` list, and it has a public constructor. So I assumed it can contain nulls. If we add a check, we can remove the `?.`.

eerhardt commented on csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs:
```diff
@@ -109,20 +109,19 @@ public Builder<T> Clear()
+                var memoryOwner = memoryAllocator.Allocate(length);
 
-                Memory.CopyTo(memory);
+                if (memoryOwner != null)
```
`memoryOwner` will be `null` when you pass in a `length == 0` to `memoryAllocator.Allocate(length)`.

Looking at this code more - it is using `_buffer.Length` which isn't correct. It should be using `Length`, which can be different (for example if I used the builder to create a `1000` elements, and then called `Clear()` and added `10` elements, I don't need a buffer big enough for `1000` elements.

eerhardt replied:
Fixing this is a little bigger than I thought it would be. So I separated it into a commit on its own.

adamsitnik replied:
Thank you for the explanation, it was not obvious to me.

Have you considered using the Null Object Pattern and returning an empty memory instead of null in such case?

eerhardt replied:
Thanks for the suggestion, @adamsitnik. I've opened https://issues.apache.org/jira/browse/ARROW-5277 to implement this.

eerhardt commented on csharp/src/Apache.Arrow/Arrays/Array.cs:
```diff
@@ -59,5 +59,19 @@ internal static void Accept<T>(T array, IArrowArrayVisitor visitor)
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
`Array` has a `protected` constructor, which means anyone can derive from it.

Also - we are overriding the virtual Dispose in `ListArray.cs` to dispose its inner `Values` array.

---
In reply to: [281211354](https://github.com/apache/arrow/pull/3925#discussion_r281211354) [](ancestors = 281211354)

eerhardt commented on csharp/src/Apache.Arrow/Memory/NativeMemoryAllocator.cs:
```diff
@@ -14,16 +14,17 @@
-        protected override Memory<byte> AllocateInternal(int length, out int bytesAllocated)
+        protected override IMemoryOwner<byte> AllocateInternal(int length, out int bytesAllocated)
         {
             // TODO: Ensure memory is released if exception occurs.
```
I think this TODO was referring to when an exception happens during this method (like if `GC.AddMemoryPressure` fails).

eerhardt commented on csharp/src/Apache.Arrow/Arrays/ArrayData.cs:
```diff
@@ -53,5 +54,24 @@ public sealed class ArrayData
+                }
+            }
+
+            if (Children != null)
```
It's just used in these 2 places, is it really worth an extra method? Also, in one case it can have nulls and the other it can't (because `ArrowBuffer` is a `struct`).

adamsitnik replied:
In 3, the nulls can be handled too with `?` (the generic implementation for struct should remove the null check) 

https://github.com/apache/arrow/blob/37a9b187276a5a68760becb816faeb0d44b00ac2/csharp/src/Apache.Arrow/RecordBatch.cs#L56-L58

eerhardt commented on csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs:
```diff
@@ -62,7 +65,7 @@ protected override async ValueTask ReadSchemaAsync()
             await ValidateFileAsync().ConfigureAwait(false);
 
             int footerLength = 0;
-            await Buffers.RentReturnAsync(4, async (buffer) =>
```
If this is shown to be a problem, we can fix it in another change.

eerhardt commented on csharp/src/Apache.Arrow/ArrowBuffer.cs:
```diff
@@ -40,16 +56,21 @@ public ReadOnlySpan<byte> Span
+                .Build(allocator);
         }
 
         public bool Equals(ArrowBuffer other)
```
I wonder if we should either remove `IEquatable` all together, or if we should delegate to the underlying `Memory`/`IMemoryOwner` for these instead. Comparing (and hashing) all the values in the buffer doesn't seem correct.

`Memory<T>` doesn't compare the values it points to, but instead checks if the underlying object/index/length are the same.

https://source.dot.net/#System.Private.CoreLib/shared/System/Memory.cs,500

eerhardt replied:
@chutchinson - do you know why `ArrowBuffer` implements `IEquatable`? Is it a necessity?

eerhardt replied:
I've opened https://issues.apache.org/jira/browse/ARROW-5278 to address this.

adamsitnik commented on csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs:
```diff
@@ -241,7 +244,7 @@ private async ValueTask ValidateMagicAsync()
             try
             {
-                await Buffers.RentReturnAsync(magicLength, async (buffer) =>
+                await ArrayPool<byte>.Shared.RentReturnAsync(magicLength, async (buffer) =>
```
@chutchinson thanks for the explanation, cool name btw ;)
